doc/configuration.md
# RouteDNS Configuration Guide

## Table of contents

- [Overview](#overview)
  - [Split Configuration](#split-configuration)
  - [Regex Formatting](https://github.com/google/re2/wiki/Syntax)
- [Listeners](#listeners)
  - [Plain DNS](#plain-dns)
  - [DNS-over-TLS](#dns-over-tls)
  - [DNS-over-HTTPS](#dns-over-https)
  - [DNS-over-DTLS](#dns-over-dtls)
  - [DNS-over-QUIC](#dns-over-quic)
  - [Admin](#admin)
- [Modifiers, Groups and Routers](#modifiers-groups-and-routers)
  - [Cache](#cache)
  - [TTL Modifier](#ttl-modifier)
  - [Round-Robin group](#round-robin-group)
  - [Fail-Rotate group](#fail-rotate-group)
  - [Fail-Back group](#fail-back-group)
  - [Random group](#random-group)
  - [Fastest group](#fastest-group)
  - [Replace](#replace)
  - [Query Blocklist](#query-blocklist)
  - [Response Blocklist](#response-blocklist)
  - [Client Blocklist](#client-blocklist)
  - [EDNS0 Client Subnet modifier](#edns0-client-subnet-modifier)
  - [EDNS0 modifier](#edns0-modifier)
  - [Static Responder](#static-responder)
  - [Static Template Responder](#static-template-responder)
  - [Drop](#drop)
  - [Response Minimizer](#response-minimizer)
  - [Response Collapse](#response-collapse)
  - [Router](#router)
  - [Rate Limiter](#rate-limiter)
  - [Fastest TCP Probe](#fastest-tcp-probe)
  - [Retrying Truncated Responses](#retrying-truncated-responses)
  - [Request Deduplication](#request-deduplication)
  - [Syslog](#syslog)
- [Resolvers](#resolvers)
  - [Plain DNS](#plain-dns-resolver)
  - [DNS-over-TLS](#dns-over-tls-resolver)
  - [DNS-over-HTTPS](#dns-over-https-resolver)
  - [Oblivious DNS (ODoH)](#oblivious-DNS-ODoH)
  - [DNS-over-DTLS](#dns-over-dtls-resolver)
  - [DNS-over-QUIC](#dns-over-quic-resolver)
  - [Bootstrap Resolver](#bootstrap-resolver)
  - [SOCKS5 Proxy Support](#socks5-proxy-support)
- [Templates](#templates)

## Overview

RouteDNS uses a config file in [TOML](https://github.com/toml-lang/toml) format which is passed to the tool as argument on the command line. The configuration is broken up into sections, each of which can contain objects. Each element has a unique identifier (name) which is used to reference it from other objects in order to build a processing pipeline. A configuration can define elements in the following sections, in any order.

- `listeners` - [Listeners](#Listeners) are effectively DNS servers that receive queries from clients and form the starting point of a pipeline. Listeners are available for several different DNS protocols.
- `routers` - [Routers](#Router) can split a pipeline into multiple processing paths based on query properties such as name, type, or client information.
- `groups` - [Groups](#Modifiers-Groups-and-Routers) contain a range of failover and load-balancing algorithms as well as elements that modify queries or responses.
- `resolvers` - [Resolvers](#Resolvers) forward queries to upstream resolvers. They are in effect DNS client implementations that connect to other servers using a variety of protocols.

Not all of these are required to make a working configuration. A most basic configuration could contain a listener (receiver) and a resolver (sender) which would be a simple proxy. The listener and the resolver could use different protocols, making this proxy also a converter.

A more complex configuration could contain multiple listeners in different protocols, a router, several modifiers, and passing queries to multiple resolvers upstream, forming a pipeline like UDP listener -> router -> cache -> DoT resolver. A single configuration can hold more than one independent pipeline.

Below an example configuration that provides two local listeners, one for plain UDP, one for plain TCP. Each query passes through a router which splits the processing into 2 paths. One path for the client 192.168.1.123, and one for the rest. Queries from 192.168.1.123 are sent through a blocklist that filters out undesirable content before getting passed to the cleanbrowsing resolver using DNS-over-TLS while everyone else will get queries answered by Cloudflare unfiltered (also using DNS-over-TLS).

```toml
[resolvers.cleanbrowsing-dot]
address = "family-filter-dns.cleanbrowsing.org:853"
protocol = "dot"

[resolvers.cloudflare-dot]
address = "1.1.1.1:853"
protocol = "dot"

[groups.cleanbrowsing-filtered]
type = "blocklist-v2"
resolvers = ["cleanbrowsing-dot"]
blocklist = [
  '.evil.com',
  '.no-good.com',
]

[routers.router1]
routes = [
  { source = "192.168.1.123/32", resolver="cleanbrowsing-filtered" },
  { resolver="cloudflare-dot" },
]

[listeners.local-udp]
address = ":53"
protocol = "udp"
resolver = "router1"

[listeners.local-tcp]
address = ":53"
protocol = "tcp"
resolver = "router1"
```

More modifiers, groups or routers can be added to the pipeline (in any order). Objects reference each other by their identifiers which have to be unique in a given configuration.

### Split Configuration

Configuration can be broken up into individual files to support large or generated configurations. Split configuration files are passed as arguments to the application:

```text
routedns example-config/split-config/*.toml
```

The same constraints on unique identifiers apply in a split configuration. The individual files are effectively concatenated prior to being loaded.

Example [split-config](../cmd/routedns/example-config/split-config).

## Listeners

Listers are query receivers that form the start of a query pipeline. Queries received by a listener are then forwarded to routers, groups, or to resolvers directly. Several DNS protocols are supported.

While nothing in the configuration references a listener (since it's the first element in a pipeline), it still requires a name that is defined like so `[listeners.NAME]`.

Common options for all listeners:

- `address` - Listen address.
- `protocol` - The DNS protocol used to receive queries, can be `udp`, `tcp`, `dot`, `doh`, `doq`.
- `resolver` - Name/identifier of the next element in the pipeline. Can be a router, group, modifier or resolver.
- `allowed-net` - Array of network addresses that are allowed to send queries to this listener, in CIDR notation, such as `["192.167.1.0/24", "::1/128"]`. If not set, no filter is applied, all clients can send queries.

Secure listeners, such as DNS-over-TLS, DNS-over-HTTPS, DNS-over-DTLS, DNS-over-QUIC and Admin support additional options to configure certificate, keys and peer validation

- `server-crt` - Server certificate file. Required.
- `server-key` - Server key file. Required.
- `ca` - CA to validate client certificated. Optional. Uses the operating system's CA store by default.
- `mutual-tls` - Requires clients to send valid (as per `ca` option) certificates before establishing a connection. Optional.

The DNS-over-HTTPS listener also accepts the client IP address from trusted reverse proxies in a particular subnet. X-Forwarded-For headers are only used if they are provided from this subnet

- `trusted-proxy` - CIDR address of trusted reverse proxy. Optional.

### Plain DNS

Regular (insecure) DNS protocol over port 53, UDP and TCP. Setting `protocol` to `udp` will start a UDP listener, and `tcp` starts a TCP listener. In many cases both are present in a configuration if RouteDNS is used to provide DNS to local services over the loopback device.

Examples:

```toml
[listeners.local-udp]
address = "127.0.0.1:53"
protocol = "udp"
resolver = "router1"

[listeners.local-tcp]
address = "127.0.0.1:53"
protocol = "tcp"
resolver = "router1"
```

### DNS-over-TLS

DNS protocol using a TLS connection (DoT) as per [RFC7858](https://tools.ietf.org/html/rfc7858). Listeners are configured with `protocol = "dot"`.

Examples:

DoT listener accepting any client. Does not require client certificates and does not validate client certificates with a CA.

```toml
[listeners.local-dot]
address = ":853"
protocol = "dot"
resolver = "cloudflare-dot"
server-crt = "/path/to/server.crt"
server-key = "/path/to/server.key"
```

DoT listener enforcing mTLS and verifying client certificates with a CA.

```toml
[listeners.local-dot]
address = ":853"
protocol = "dot"
resolver = "cloudflare-dot"
server-crt = "/path/to/server.crt"
server-key = "/path/to/server.key"
ca = "/path/to/ca.crt"
mutual-tls = true
```

Example config files: [mutual-tls-dot-server.toml](../cmd/routedns/example-config/mutual-tls-dot-server.toml)

### DNS-over-HTTPS

As per [RFC8484](https://tools.ietf.org/html/rfc8484), DNS using the HTTPS protocol are configured with `protocol = "doh"`. By default, DoH uses TCP as transport, but it can also be run over QUIC by providing the option `transport = "quic"`. For TCP transport, TLS can be disabled with the `no-tls = true` option which can be used for testing or when the server is only accessible via reverse proxy that terminates TLS already.

Examples:

DoH listener accepting queries from any client.

```toml
[listeners.local-doh]
address = ":443"
protocol = "doh"
resolver = "cloudflare-dot"
server-crt = "/path/to/server.crt"
server-key = "/path/to/server.key"
```

DoH over QUIC listener.

```toml
[listeners.local-doh-quic]
address = ":1443"
protocol = "doh"
transport = "quic"
resolver = "cloudflare-dot"
server-crt = "example-config/server.crt"
server-key = "example-config/server.key"
```

DoH behind a reverse proxy. Clients are expected to connect to a reverse proxy in this subnet, which will provide their IP address in the X-Forwarded-For header. RouteDNS will trust this header from proxies in the subnet listed in `trusted-proxy`

```toml
[listeners.local-doh]
address = ":443"
protocol = "doh"
resolver = "cloudflare-dot"
server-crt = "/path/to/server.crt"
server-key = "/path/to/server.key"
frontend = { trusted-proxy = "192.168.1.0/24" }
```

Example config files: [mutual-tls-doh-server.toml](../cmd/routedns/example-config/mutual-tls-doh-server.toml), [doh-quic-server.toml](../cmd/routedns/example-config/doh-quic-server.toml), [doh-behind-proxy.toml](../cmd/routedns/example-config/doh-behind-proxy.toml), [doh-no-tls.toml](../cmd/routedns/example-config/doh-no-tls.toml)

### DNS-over-DTLS

Similar to DoT, but uses a DTLS (UDP) connection as transport as per [RFC9894](https://tools.ietf.org/html/rfc8094). Configured with `protocol = "dtls"`.

Examples:

DTLS listener.

```toml
[listeners.local-dtls]
address = ":853"
protocol = "dtls"
resolver = "cloudflare-dot"
server-crt = "example-config/server-ec.crt"
server-key = "example-config/server-ec.key"
```

Example config files: [dtls-server.toml](../cmd/routedns/example-config/dtls-server.toml)

### DNS-over-QUIC

Similar to DoT, but uses a QUIC connection as transport as per [RFC9250](https://datatracker.ietf.org/doc/rfc9250/). Configured with `protocol = "doq"`. Note that this is different from DoH over QUIC. See [DNS-over-HTTPS](#DNS-over-HTTPS) for how to configure this.

Note: Support for the QUIC protocol is still experimental. For the purpose of DNS, there are two implementations, DNS-over-QUIC ([RFC9250](https://datatracker.ietf.org/doc/rfc9250/)) as well as DNS-over-HTTPS using QUIC. Both methods are supported by RouteDNS, client and server implementations.

Examples:

DoQ listener accepting queries from all clients.

```toml
[listeners.local-doq]
address = ":8853"
protocol = "doq"
resolver = "cloudflare-dot"
server-crt = "example-config/server.crt"
server-key = "example-config/server.key"
```

Example config files: [doq-listener.toml](../cmd/routedns/example-config/doq-listener.toml)

### Oblivious DNS over HTTPS (ODoH)

ODoH ([RFC9230](https://datatracker.ietf.org/doc/rfc9230/)) 

Providing the key-seed is not necessary if the listener is running in proxy mode. If it is not set, the listener will generate a new random key on each launch. A key seed can be manually generated using for example openssl: `openssl rand -hex 16`

By default the ODoH listener listens on `/proxy` and `/dns-query`. Additionally, it will host its HPKE config under `/.well-known/odohconfigs`. If `odoh-mode  = "proxy"` is set, it will only listen and handle ODoH proxy requests on `/proxy`. If set to target the listener will only handle the ODoH queries on `/dns-query`.

```toml
[listeners.local-odoh]
address = ":443"
protocol = "odoh"
resolver = "cloudflare"

# The key seed is used to generate the HPKE keypair. 
key-seed = "414dd55667a0cdff72dfbbd8515a9e0a"
# odoh-mode allows restricting to only "proxy" mode or "target" mode. Default or left empty enables both modes.
odoh-mode  = "target"

# If enabled, the listener will also respond to regular DoH queries using the same resolver. When not set or false, DoH queries are ignored. Has no effect if odoh-mode is set to proxy
allow-doh = true

# TLS information
server-crt = "example-config/server.crt"
server-key = "example-config/server.key"
```

### Admin

The Admin listener provides metrics on RouteDNS usage and performance at https://{address}/routedns/vars/ in [expvar](https://pkg.go.dev/expvar) format. These metrics can be exported to be usable by Prometheus using [prometheus-expvar-exporter](https://github.com/albertito/prometheus-expvar-exporter). An example configuration is provided below.

Examples:

```toml
[listeners.local-admin]
address = "127.0.0.7:443"
protocol = "admin"
server-crt = "example-config/server.crt"
server-key = "example-config/server.key"
```

Example config files: [admin.toml](../cmd/routedns/example-config/admin.toml), [prometheus-exporter](../cmd/routedns/example-config/prometheus-exporter/)

## Modifiers, Groups and Routers

### Cache

A cache will store the responses to queries in memory and respond to further identical queries with the same response. To determine how long an item is kept in memory, the cache uses the lowest TTL of the RRs in the response. Responses served from the cache have their TTL updated according to the time the records spent in memory. If a query has an [ECS Subnet](https://tools.ietf.org/html/rfc7871) option, the subnet address forms part of they key to support subnet-specific answers.

Caches can be combined with a [TTL Modifier](#TTL-Modifier) to avoid too many cache-misses due to excessively low TTL values.

It is possible to pre-define a query name that will flush the cache if received from a client.

The content of memory caches can be persisted to and loaded from disk.

#### Configuration

Caches are instantiated with `type = "cache"` in the groups section of the configuration.

Options:

- `resolvers` - Array of upstream resolvers, only one is supported.
- `cache-size` - Max number of responses to cache. Defaults to 0 which means no limit. Deprecated, set limit in the backend instead.
- `cache-negative-ttl` - TTL (in seconds) to apply to responses without a SOA. Default: 60. Optional
- `cache-rcode-max-ttl` - Map of RCODE to max TTL (in seconds) to use for records based on the status code regardless of SOA. Response codes are given in their numerical form: 0 = NOERROR, 1 = FORMERR, 2 = SERVFAIL, 3 = NXDOMAIN, ... See [rfc2929#section-2.3](https://tools.ietf.org/html/rfc2929#section-2.3) for a more complete list. For example `{1 = 60, 3 = 60}` would set a limit on how long FORMERR or NXDOMAIN responses can be cached.
- `cache-answer-shuffle` - Specifies a method for changing the order of cached A/AAAA answer records. Possible values `random` or `round-robin`. Defaults to static responses if not set.
- `cache-harden-below-nxdomain` - Return NXDOMAIN for domain queries if the parent domain has a cached NXDOMAIN. See [RFC8020](https://tools.ietf.org/html/rfc8020).
- `cache-flush-query` - A query name (FQDN with trailing `.`) that if received from a client will trigger a cache flush (reset). Inactive if not set. Simple way to support flushing the cache by sending a pre-defined query name of any type. If successful, the response will be empty. The query will not be forwarded upstream by the cache.
- `cache-prefetch-trigger`- If a query is received for a record with less that `cache-prefetch-trigger` TTL left, the cache will send another, independent query to upstream with the goal of automatically refreshing the record in the cache with the response.
- `cache-prefetch-eligible` - Only records with at least `prefetch-eligible` seconds TTL are eligible to be prefetched.
- `backend` - Define what kind of storage is used for the cache. Contains multiple keys depending on type that can configure the behavior. Defaults to `memory` backend if not configued.

Backends:

**Memory backend**

The memory backend will keep all cache items in memory. It can be configured to write the content of the cache to disk on shutdown. Memory backend config has the following options:

- `type="memory"`
- `size` - Max number of responses to cache. Defaults to 0 which means no limit.
- `filename` - File to use for persistent storage to disk. The cache will be initialized with the content from the file and it'll write the content to the same file on shutdown. Defaults to no persistence
- `save-interval` - Interval (in seconds) to save the cache to file. Optional. If not set, the file is written only on shutdown.

**Redis backend**

The `redis` backend stores cached items in a Redis database. This allows multiple instances of routedns to share a common cache backend. The following options are supported:

- `type="redis"`
- `redis-network` - The network type, either `tcp` or `unix`. Defaults to `tcp`.
- `redis-address` - Address of redis database, host:port
- `redis-username` - Redis username
- `redis-password` - Redis password
- `redis-db` - Redis database to be selected
- `redis-key-prefix` - Prefixes the key of every record with this string. This can be used to share a database with other clients and avoid key conflicts.
- `redis-max-retries` - Maximum number of retries before giving up. Default is 3 retries; -1 (not 0) disables retries.
- `redis-min-retry-backoff` - Minimum back-off between each retry in milliseconds. Default is 8 milliseconds; -1 disables back-off.
- `redis-max-retry-backoff` - Maximum back-off between each retry in milliseconds. Default is 512 milliseconds; -1 disables back-off.

#### Examples

Simple cache without size-limit:

```toml
[groups.cloudflare-cached]
type = "cache"
resolvers = ["cloudflare-dot"]
backend = {type = "memory"}
```

Cache that only stores up to 1000 records in memory and keeps negative responses for 1h. Responses are randomized for cached responses.

```toml
[groups.cloudflare-cached]
type = "cache"
resolvers = ["cloudflare-dot"]
cache-negative-ttl = 3600
cache-answer-shuffle = "random"
backend = {type = "memory", size = 1000}
```

Cache that is flushed if a query for `flush.cache.` is received. Also persists the cache to disk.

```toml
[groups.cloudflare-cached]
type = "cache"
resolvers = ["cloudflare-dot"]
cache-flush-query = "flush.cache."
backend = {type = "memory", filename = "/var/tmp/cache.json"}
```

Cache that is uses Redis as backend.

```toml
[groups.cloudflare-cached]
type = "cache"
resolvers = ["cloudflare-dot"]
cache-flush-query = "flush.cache."
backend = {type = "redis", redis-address = "127.0.0.1:6379", redis-key-prefix = "routedns-"}
```

Example config files: [cache.toml](../cmd/routedns/example-config/cache.toml), [block-split-cache.toml](../cmd/routedns/example-config/block-split-cache.toml), [cache-flush.toml](../cmd/routedns/example-config/cache-flush.toml), [cache-with-prefetch.toml](../cmd/routedns/example-config/cache-with-prefetch.toml), [cache-rcode.toml](../cmd/routedns/example-config/cache-rcode.toml), [cache-redis.toml](../cmd/routedns/example-config/cache-redis.toml)

### TTL modifier

A TTL modifier is used to adjust the time-to-live (TTL) of DNS responses. This is used to avoid frequently making the same queries to upstream because many responses have a value that is unreasonably low as outlined in this [blog](https://blog.apnic.net/2019/11/12/stop-using-ridiculously-low-dns-ttls). It's also possible to restrict very high TTL values that might be used in DNS poisoning attacks.

The limits are applied to all RRs in a response.

#### Configuration

Caches are instantiated with `type = "ttl-modifier"` in the groups section of the configuration.

Options:

- `resolvers` - Array of upstream resolvers, only one is supported.
- `ttl-select` - Optional TTL selection function. Possible values "lowest", "highest", "average", "first", "last".
  - `lowest` - Lowest TTL of all response records.
  - `highest` - Highest TTL of all response records.
  - `average` - Average TTL of all response records.
  - `first` - First TTL.
  - `last` - Last TTL.
  - `random` - Random TTL between `ttl-min` and `ttl-max`. Note that not setting `ttl-max` will result in very high TTL values.
- `ttl-min` - TTL minimum (in seconds) to apply to responses.
- `ttl-max` - TTL minimum (in seconds) to apply to responses.

`ttl-min` and `ttl-max` are optional, but if configured define a floor/ceiling regardless of what `ttl-select` function is given.

#### Examples

TTL modifier that returns responses with TTL of between 1h and one day:

```toml
[groups.cloudflare-updated-ttl]
type = "ttl-modifier"
resolvers = ["cloudflare-dot"]
ttl-min = 3600
ttl-max = 86400
```

TTL modifier returning the average TTL of all records, with a max of 1 day.

```toml
[groups.cloudflare-updated-ttl]
type = "ttl-modifier"
resolvers = ["cloudflare-dot"]
ttl-select = "average"
ttl-max = 86400
```

Example config files: [ttl-modifier.toml](../cmd/routedns/example-config/ttl-modifier.toml), [ttl-modifier-average.toml](../cmd/routedns/example-config/ttl-modifier-average.toml)

### Round-Robin group

A Round-Robin balancer groups multiple upstream resolvers and sends every received query to the next resolver. It effectively balances the query load evenly over a number of upstream resolvers or modifiers.

#### Configuration

Round-Robin groups are instantiated with `type = "round-robin"` in the groups section of the configuration.

Options:

- `resolvers` - An array of upstream resolvers or modifiers.

#### Examples

```toml
[groups.google-udp]
resolvers = ["google-udp-8-8-8-8", "google-udp-8-8-4-4"]
type = "round-robin"
```

### Fail-Rotate group

In a Fail-Rotate group, one of the upstream resolvers or modifiers is active and receives all queries. If the active resolver fails, i.e. no response or returns SERVFAIL, the next becomes active and the request is retried. If the last resolver fails the first becomes the active again. There's no time-based automatic fail-back.

#### Configuration

Round-Robin groups are instantiated with `type = "fail-rotate"` in the groups section of the configuration.

Options:

- `resolvers` - An array of upstream resolvers or modifiers.
- `servfail-error` - If `true`, a SERVFAIL response from an upstream resolver is considered a failure triggering a switch to the next resolver. This can happen when DNSSEC validation fails for example. Default `false`.

#### Examples

```toml
[groups.google-udp]
resolvers = ["google-udp-8-8-8-8", "google-udp-8-8-4-4"]
type = "fail-rotate"
```

### Fail-Back group

Similar to [fail-rotate](#Fail-Rotate-group) but will attempt to fall back to the original order (prioritizing the first) if there are no failures for a minute. Failure means either no response or it returns SERVFAIL.

#### Configuration

Fail-Back groups are instantiated with `type = "fail-back"` in the groups section of the configuration.

Options:

- `resolvers` - An array of upstream resolvers or modifiers. The first in the array is the preferred resolver.
- `reset-after` - Time in seconds before switching from an alternative resolver back to the preferred resolver (first in the list), default 60. Note: This is not a timeout argument. After a failure of the preferred resolver, this defines the amount of time to use alternative/failover resolvers before switching back to the preferred. You can have as many resolvers in the array as the time limit allows.
- `servfail-error` - If `true`, a SERVFAIL response from an upstream resolver is considered a failure triggering a failover. This can happen when DNSSEC validation fails for example. Default `false`.

#### Examples

```toml
[groups.my-failback-group]
resolvers = ["company-dns", "cloudflare-dot"]
type = "fail-back"
```


cmd/routedns/config.go
package main

import (
	"bytes"
	"io"
	"net"
	"os"

	"github.com/BurntSushi/toml"
	rdns "github.com/folbricht/routedns"
)

type config struct {
	Title             string
	BootstrapResolver resolver `toml:"bootstrap-resolver"`
	Listeners         map[string]listener
	Resolvers         map[string]resolver
	Groups            map[string]group
	Routers           map[string]router
}

type listener struct {
	Address    string
	Protocol   string
	Transport  string
	Resolver   string
	CA         string
	ServerKey  string   `toml:"server-key"`
	ServerCrt  string   `toml:"server-crt"`
	MutualTLS  bool     `toml:"mutual-tls"`
	NoTLS      bool     `toml:"no-tls"` // Disable TLS in DoH servers
	AllowedNet []string `toml:"allowed-net"`
	KeySeed    string   `toml:"key-seed"`  // ODoH HPKE key seed, 16 byte hex key. Generate for example with: "openssl rand -hex 16"
	OdohMode   string   `toml:"odoh-mode"` // switch between ODoH mode - accepts "proxy", "target" or does both if empty/other
	AllowDoH   bool     `toml:"allow-doh"` // Allow ODoH listeners to also handle DoH queries to /dns-query
	Frontend   dohFrontend
}

// DoH listener frontend options
type dohFrontend struct {
	HTTPProxyNet string `toml:"trusted-proxy"`
}

type resolver struct {
	Address       string
	Protocol      string
	Transport     string
	DoH           doh
	CA            string
	ClientKey     string `toml:"client-key"`
	ClientCrt     string `toml:"client-crt"`
	ServerName    string `toml:"server-name"` // TLS server name presented in the server certificate
	BootstrapAddr string `toml:"bootstrap-address"`
	LocalAddr     string `toml:"local-address"`
	EDNS0UDPSize  uint16 `toml:"edns0-udp-size"` // UDP resolver option
	QueryTimeout  int    `toml:"query-timeout"`  // Query timeout in seconds

	// Proxy configuration
	Socks5Address      string `toml:"socks5-address"`
	Socks5Username     string `toml:"socks5-username"`
	Socks5Password     string `toml:"socks5-password"`
	Socks5ResolveLocal bool   `toml:"socks5-resolve-local"` // Resolve DNS server address locally (i.e. bootstrap-resolver), not on the SOCK5 proxy

	//QUIC and DoH/3 configuration
	Use0RTT bool `toml:"enable-0rtt"`
	// URL for Oblivious DNS target
	Target       string `toml:"target"`
	TargetConfig string `toml:"target-config"`
}

// DoH-specific resolver options
type doh struct {
	Method string
}

// Cache backend options
type cacheBackend struct {
	Type                 string // Cache backend type.Defaults to "memory"
	Size                 int    // Max number of items to keep in the cache. Default 0 == unlimited. Deprecated, use backend
	GCPeriod             int    `toml:"gc-period"` // Time-period (seconds) used to expire cached items
	Filename             string // File to load/store cache content, optional, for "memory" type cache
	SaveInterval         int    `toml:"save-interval"`           // Seconds to write the cache to file
	RedisNetwork         string `toml:"redis-network"`           // The network type, either tcp or unix. Defaults to tcp.
	RedisAddress         string `toml:"redis-address"`           // Address for redis cache
	RedisUsername        string `toml:"redis-username"`          // Redis username
	RedisPassword        string `toml:"redis-password"`          // Redis password
	RedisDB              int    `toml:"redis-db"`                // Redis database to be selected after connecting to the server
	RedisKeyPrefix       string `toml:"redis-key-prefix"`        // Prefix any cache entry
	RedisMaxRetries      int    `toml:"redis-max-retries"`       // Maximum number of retries before giving up. Default is 3 retries; -1 (not 0) disables retries.
	RedisMinRetryBackoff int    `toml:"redis-min-retry-backoff"` // Minimum back-off between each retry. Default is 8 milliseconds; -1 disables back-off.
	RedisMaxRetryBackoff int    `toml:"redis-max-retry-backoff"` // Maximum back-off between each retry. Default is 512 milliseconds; -1 disables back-off.
}

type group struct {
	Resolvers  []string
	Type       string
	Replace    []rdns.ReplaceOperation // only used by "replace" type
	ECSOp      string                  `toml:"ecs-op"`      // ECS modifier operation, "add", "delete", "privacy"
	ECSAddress net.IP                  `toml:"ecs-address"` // ECS address. If empty for "add", uses the client IP. Ignored for "privacy" and "delete"
	ECSPrefix4 uint8                   `toml:"ecs-prefix4"` // ECS IPv4 address prefix, 0-32. Used for "add" and "privacy"
	ECSPrefix6 uint8                   `toml:"ecs-prefix6"` // ECS IPv6 address prefix, 0-128. Used for "add" and "privacy"
	TTLMin     uint32                  `toml:"ttl-min"`     // TTL minimum to apply to responses in the TTL-modifier
	TTLMax     uint32                  `toml:"ttl-max"`     // TTL maximum to apply to responses in the TTL-modifier
	TTLSelect  string                  `toml:"ttl-select"`  // Modifier selection function, "lowest", "highest", "average", "first", "last", "random"
	EDNS0Op    string                  `toml:"edns0-op"`    // EDNS0 modifier operation, "add" or "delete"
	EDNS0Code  uint16                  `toml:"edns0-code"`  // EDNS0 modifier option code
	EDNS0Data  []byte                  `toml:"edns0-data"`  // EDNS0 modifier option data

	// Failover/Failback options
	ResetAfter    int  `toml:"reset-after"`    // Time in seconds after which to reset resolvers in fail-back and random groups, default 60.
	ServfailError bool `toml:"servfail-error"` // If true, SERVFAIL responses are considered errors and cause failover etc.

	// Cache options
	Backend                  *cacheBackend
	GCPeriod                 int               `toml:"gc-period"`                   // Time-period (seconds) used to expire cached items in the "cache" type. Deprecated, use backend
	CacheSize                int               `toml:"cache-size"`                  // Max number of items to keep in the cache. Default 0 == unlimited. Deprecated, use backend
	CacheNegativeTTL         uint32            `toml:"cache-negative-ttl"`          // TTL to apply to negative responses, default 60.
	CacheAnswerShuffle       string            `toml:"cache-answer-shuffle"`        // Algorithm to use for modifying the response order of cached items
	CacheHardenBelowNXDOMAIN bool              `toml:"cache-harden-below-nxdomain"` // Return NXDOMAIN if an NXDOMAIN is cached for a parent domain
	CacheFlushQuery          string            `toml:"cache-flush-query"`           // Flush the cache when a query for this name is received
	PrefetchTrigger          uint32            `toml:"cache-prefetch-trigger"`      // Prefetch when the TTL of a query has fallen below this value
	PrefetchEligible         uint32            `toml:"cache-prefetch-eligible"`     // Only records with TTL greater than this are considered for prefetch
	CacheRcodeMaxTTL         map[string]uint32 `toml:"cache-rcode-max-ttl"`         // Rcode specific max TTL to keep in the cache

	// Blocklist options
	Blocklist []string // Blocklist rules, only used by "blocklist" type
	Format    string   // Blocklist input format: "regex", "domain", or "hosts"
	Source    string   // Location of external blocklist, can be a local path or remote URL
	Refresh   int      // Blocklist refresh when using an external source, in seconds

	// Blocklist-v2 options
	Filter            bool     // Filter response records rather than return NXDOMAIN
	BlockListResolver string   `toml:"blocklist-resolver"`
	AllowListResolver string   `toml:"allowlist-resolver"`
	BlocklistFormat   string   `toml:"blocklist-format"` // only used for static blocklists in the config
	BlocklistSource   []list   `toml:"blocklist-source"`
	BlocklistRefresh  int      `toml:"blocklist-refresh"`
	Allowlist         []string // Rules to override the blocklist rules
	AllowlistFormat   string   `toml:"allowlist-format"` // only used for static allowlists in the config
	AllowlistSource   []list   `toml:"allowlist-source"`
	AllowlistRefresh  int      `toml:"allowlist-refresh"`
	LocationDB        string   `toml:"location-db"` // GeoIP database file for response blocklist. Default "/usr/share/GeoIP/GeoLite2-City.mmdb"
	Inverted          bool     // Only allow IPs on the blocklist. Supported in response-blocklist-ip and response-blocklist-name

	// Static responder options
	Answer   []string
	NS       []string
	Extra    []string
	RCode    int
	EDNS0EDE struct {
		Code uint16 `toml:"code"` // Code defined in https://datatracker.ietf.org/doc/html/rfc8914
		Text string `toml:"text"` // Extra text containing additional information
	} `toml:"edns0-ede"` // Extended DNS Errors
	Truncate bool `toml:"truncate"` // When true, TC-Bit is set

	// Rate-limiting options
	Requests      uint   // Number of requests allowed
	Window        uint   // Time period in seconds for the requests
	Prefix4       uint8  // Prefix bits to identify IPv4 client
	Prefix6       uint8  // Prefix bits to identify IPv6 client
	LimitResolver string `toml:"limit-resolver"` // Resolver to use when rate-limit exceeded

	// Fastest-TCP probe options
	Port          int
	WaitAll       bool   `toml:"wait-all"`        // Wait for all probes to return and respond with a sorted list. Generally slower
	SuccessTTLMin uint32 `toml:"success-ttl-min"` // Set the TTL of records that were probed successfully

	// Response Collapse options
	NullRCode int `toml:"null-rcode"` // Response code if after collapsing, no answers are left

	// Truncate-Retry options
	RetryResolver string `toml:"retry-resolver"`

	// Syslog options
	Network     string `toml:"network"`  // "udp", "tcp", "unix"
	Address     string `toml:"address"`  // Endpoint address, defaults to local syslog server
	Priority    string `toml:"priority"` // Syslog priority, "emergency", "alert", "critical", "error", "warning", "notice", "info", "debug"
	Tag         string `toml:"tag"`
	LogRequest  bool   `toml:"log-request"`  // Logs request records to syslog
	LogResponse bool   `toml:"log-response"` // Logs response records to syslog
	Verbose     bool   `toml:"verbose"`      // When logging responses, include types that don't match the query type
}

// Block/Allowlist items for blocklist-v2
type list struct {
	Name         string
	Format       string
	Source       string
	CacheDir     string `toml:"cache-dir"`     // Where to store copies of remote blocklists for faster startup
	AllowFailure bool   `toml:"allow-failure"` // Don't fail on error and keep using the prior ruleset
}

type router struct {
	Routes []route
}

type route struct {
	Type          string // Deprecated, use "Types" instead
	Types         []string
	Class         string
	Name          string
	Source        string
	Weekdays      []string // 'mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun'
	After, Before string   // Hour:Minute in 24h format, for example "14:30"
	Invert        bool     // Invert the result of the match
	DoHPath       string   `toml:"doh-path"` // DoH query path if received over DoH (regexp)
	Resolver      string
	Listener      string // ID of the listener that received the original request
	TLSServerName string `toml:"servername"` // TLS servername
}

// LoadConfig reads a config file and returns the decoded structure.
func loadConfig(name ...string) (config, error) {
	b := new(bytes.Buffer)
	var c config
	for _, fn := range name {
		if err := loadFile(b, fn); err != nil {
			return c, err
		}
		b.WriteString("\n")
	}
	_, err := toml.DecodeReader(b, &c)
	return c, err
}

func loadFile(w io.Writer, name string) error {
	f, err := os.Open(name)
	if err != nil {
		return err
	}
	defer f.Close()
	_, err = io.Copy(w, f)
	return err
}

func parseCIDRList(networks []string) ([]*net.IPNet, error) {
	var out []*net.IPNet
	for _, s := range networks {
		_, n, err := net.ParseCIDR(s)
		if err != nil {
			return out, err
		}
		out = append(out, n)
	}
	return out, nil
}
